src/box/replication.h

```diff
@@ -401,6 +401,8 @@ struct replica {
 	struct applier *applier;
 	/** Relay thread. */
 	struct relay *relay;
+	/** Relay final join thread */
+	struct relay *relay_join;
```

Comment:
IMO storing the "join relay" separately from "normal relay" looks rather ugly. I'd suggest to resort to this hack only if reusing the normal "relay" field didn't work for some reason.

But it works: I tried the following dirty patch, and everything seems fine. Be careful when applying: you have to find a better way for determining that relay is joining (along the lines where `relay->replica == NULL` is replaced with `relay->stream.yield != relay_yield_and_send_heartbeat`)
<details><summary>diff</summary>

```diff
diff --git a/src/box/box.cc b/src/box/box.cc
index 4a0536f7b..4dea8b62e 100644
--- a/src/box/box.cc
+++ b/src/box/box.cc
@@ -4203,6 +4203,13 @@ box_process_register(struct iostream *io, const struct xrow_header *header)
 			  "registration of non-anonymous nodes.");
 	}
 
+	/* Don't allow multiple relays for the same replica */
+	if (replica != NULL &&
+	    relay_get_state(replica->relay) == RELAY_FOLLOW) {
+		tnt_raise(ClientError, ER_CFG, "replication",
+			  "duplicate connection with the same replica UUID");
+	}
+
 	/* See box_process_join() */
 	box_check_writable_xc();
 	struct space *space = space_cache_find_xc(BOX_CLUSTER_ID);
@@ -4240,7 +4247,7 @@ box_process_register(struct iostream *io, const struct xrow_header *header)
 	 * (req.vclock, stop_vclock) so that it gets its
 	 * registration.
 	 */
-	relay_final_join(io, header->sync, &req.vclock, &stop_vclock);
+	relay_final_join(replica, io, header->sync, &req.vclock, &stop_vclock);
 	say_info("final data sent.");
 
 	RegionGuard region_guard(&fiber()->gc);
@@ -4403,7 +4410,8 @@ box_process_join(struct iostream *io, const struct xrow_header *header)
 	 * Final stage: feed replica with WALs in range
 	 * (start_vclock, stop_vclock).
 	 */
-	relay_final_join(io, header->sync, &start_vclock, &stop_vclock);
+	relay_final_join(replica, io, header->sync, &start_vclock,
+			 &stop_vclock);
 	say_info("final data sent.");
 
 	/* Send end of WAL stream marker */
diff --git a/src/box/relay.cc b/src/box/relay.cc
index a231018d6..27b43a59c 100644
--- a/src/box/relay.cc
+++ b/src/box/relay.cc
@@ -526,17 +526,20 @@ relay_final_join_f(va_list ap)
 }
 
 void
-relay_final_join(struct iostream *io, uint64_t sync,
+relay_final_join(struct replica *replica, struct iostream *io, uint64_t sync,
 		 struct vclock *start_vclock, struct vclock *stop_vclock)
 {
-	struct relay *relay = relay_new(NULL);
-	if (relay == NULL)
-		diag_raise();
-
+	struct relay *relay = replica->relay;
+	assert(relay->state != RELAY_FOLLOW);
+	/*
+	 * As a new thread is started for the final join stage, its cancellation
+	 * should be handled properly during unexpected shutdown, so, the
+	 * relay's pointer is saved in the replica structure and the thread is
+	 * cancelled during replication_free().
+	 */
 	relay_start(relay, io, sync, relay_send_row, relay_yield, UINT64_MAX);
 	auto relay_guard = make_scoped_guard([=] {
 		relay_stop(relay);
-		relay_delete(relay);
 	});
 	/*
 	 * Save the first vclock as 'received'. Because it was really received.
@@ -1212,7 +1215,7 @@ relay_send_row(struct xstream *stream, struct xrow_header *packet)
 {
 	struct relay *relay = container_of(stream, struct relay, stream);
 	/* Do not send heartbeats during a final join. */
-	if (relay->replica != NULL)
+	if (relay->stream.yield == relay_yield_and_send_heartbeat)
 		relay_send_heartbeat_on_timeout(relay);
 	if (packet->group_id == GROUP_LOCAL) {
 		/*
@@ -1247,7 +1250,7 @@ relay_send_row(struct xstream *stream, struct xrow_header *packet)
 	 * it). In the latter case packet's LSN is less than or equal to
 	 * local master's LSN at the moment it received 'SUBSCRIBE' request.
 	 */
-	if (relay->replica == NULL ||
+	if (relay->stream.yield != relay_yield_and_send_heartbeat ||
 	    packet->replica_id != relay->replica->id ||
 	    packet->lsn <= vclock_get(&relay->local_vclock_at_subscribe,
 				      packet->replica_id)) {
diff --git a/src/box/relay.h b/src/box/relay.h
index 733d6bc31..5bbc87e46 100644
--- a/src/box/relay.h
+++ b/src/box/relay.h
@@ -138,7 +138,7 @@ relay_initial_join(struct iostream *io, uint64_t sync, struct vclock *vclock,
  * @param sync      sync from incoming JOIN request
  */
 void
-relay_final_join(struct iostream *io, uint64_t sync,
+relay_final_join(struct replica *replica, struct iostream *io, uint64_t sync,
 		 struct vclock *start_vclock, struct vclock *stop_vclock);
 
 /**
```
</details>

Reply:
Reusing the same relay was am initial idea, however, it didn't work, as I didn't came with idea of fully prohibiting register while the relay is still alive.